src/poetry/core/semver/version_range_constraint.py

```diff
@@ -36,44 +31,78 @@ def include_min(self) -> bool:
     def include_max(self) -> bool:
         raise NotImplementedError()
 
-    def allows_lower(self, other: VersionRangeConstraint) -> bool:
+    @property
+    def allowed_min(self) -> Version | None:
         if self.min is None:
-            return other.min is not None
+            return None
+
+        if not self.include_min or self.min.is_unstable():
+            return self.min
+
+        if self.min == self.max and (self.include_min or self.include_max):
+            # this is an equality range
+            return self.min
 
-        if other.min is None:
+        return self.min
+
+    @property
+    def allowed_max(self) -> Version | None:
+        if self.max is None:
+            return None
+
+        if self.include_max or self.max.is_unstable():
+            return self.max
+
+        if self.min == self.max and (self.include_min or self.include_max):
+            # this is an equality range
+            return self.max
+
+        return self.max.next_devrelease()
```

Comment:
As already mentioned before in another place, this is absolutely confusing. Should really be `first_devrelease()` or similar.